05.asynchronous_sqlite3/promise.js

```diff
@@ -0,0 +1,54 @@
+import sqlite3 from "sqlite3";
+
+const db = new sqlite3.Database(":memory:");
+
+function run(db, sql, bindValue) {
+  return new Promise((resolve) => {
+    db.run(sql, [bindValue], function () {
+      resolve(this);
+    });
+  });
+}
+
+function each(db, sql) {
+  return new Promise((resolve) => {
+    const responses = [];
+    db.each(sql, (err, response) => {
+      responses.push(response);
+      resolve(responses);
+    });
+  });
+}
+
+function close(db) {
+  return new Promise((resolve) => {
+    db.close();
+    resolve();
+  });
+}
+
+run(db, "CREATE TABLE books (id INTEGER PRIMARY KEY, title TEXT)")
+  .then(() => run(db, "INSERT INTO books (title) VALUES (?)", "title1"))
+  .then((record) => {
+    console.log(`lastID: ${record.lastID}`);
+  })
+  .then(() => run(db, "INSERT INTO books (title) VALUES (?)", "title2"))
```

Comment:
console.log の実行は非同期ではないので、promise として待つ必要はないと多みます。次の run の処理を console.log を実行している関数の中にまとめられそうです。

Reply:
> console.log の実行は非同期ではないので、promise として待つ必要はないと思います。

`record.lastID`を`function () { resolve(this); }`で取得しているので、.thenで繋ぐ必要があると思っていたのですが、他の方法がわかりませんでした。ご教示お願いいたします。

node-sqlite3 API
you must use an old-school function () { ... } style callback rather than a lambda function, otherwise this.lastID and this.changes will be undefined.

> 次の run の処理を console.log を実行している関数の中にまとめられそうです。

このような意味でしょうか？
【試したこと】
```
run(db, "CREATE TABLE books (id INTEGER PRIMARY KEY, title TEXT, content TEXT)")
  .then(() =>
    run(
      db,
      "INSERT INTO books (title, content) VALUES (?, ?)",
      "title1",
      "content1"
    )
  )
  .then((record) => {
    console.log(`lastID: ${record.lastID}`);
    run(
      db,
      "INSERT INTO books (title, content) VALUES (?, ?)",
      "title2",
      "content2"
    );
  })
  .then((record) => {
    console.log(`lastID: ${record.lastID}`);
    run(
      db,
      "INSERT INTO books (title, content) VALUES (?, ?)",
      "title3",
      "content3"
    );
  })
  .then((record) => {
    console.log(`lastID: ${record.lastID}`);
    run(
      db,
      "INSERT INTO books (title, content) VALUES (?, ?)",
      "title4",
      "content4"
    );
  })
  ```

Reply:
> このような意味でしょうか？

はい、そうです。1 点だけ修正が必要な箇所があると思うので、自分で修正してみてください。

Reply:
ヒントありがとうございます。
returnを付ければ良いのですね。修正しました。